src/Discovery/Walk.hs

```diff
@@ -48,5 +54,21 @@ walk f = walkDir $ \dir subdirs files -> do
     WalkSkipAll -> pure $ WalkExclude subdirs
     WalkStop -> pure WalkFinish
 
+walk' ::
+  forall m o.
+  (MonadIO m, Monoid o) =>
+  (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m (o, WalkStep)) ->
```

Comment:
The type here is awkward for sure.

The difficulty is that we need to keep the existing `walk`, because it's still useful for more-complex discovery (see: Maven.Pom), so modifying the existing `WalkStep` datatype to add another parameter is something we probably shouldn't do -- and confusingly adding an additional `DiscoveryStep` type that's only used for one type of file walking is something I'd like to avoid. (I do like the smart constructors there though! That would be a good design if we went down that path)

I do think that the type signature of `walk'` is informative though, specifically the `Monoid` constraint:

```hs
walk' ::
  (MonadIO m, Monoid o) =>
  (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m (o, WalkStep)) ->
  Path Abs Dir ->
  m o
```

notably, `walk'` returns `m o`, and the only thing we know about `o` is that it has a `Monoid` instance. so in the degenerate case, this is just `pure mempty`.

in the interesting case, because the "walkstep function" returns `o`s, the only interesting thing we can do with the `o`s is to smash them together (`<>`), and return the result of walking the filetree, smashing together the `o`s (`m o`)

Reply:
If you're against adding a new `DiscoveryStep` type, then let's create smart constructors for the existing `(o, WalkStep)` tuple.  This allows us a smoother API into the admittedly, yet unavoidably, awkward type.

```haskell
discoveryContinue :: Monoid o => o -> (o, WalkStep)
discoveryContinue o = (o, WalkContinue)
```